### PR TITLE
Support files with missing tail header

### PR DESCRIPTION
**Pull request by udoprog**
I've [found a number of archives out in the wild](https://github.com/udoprog/kernelstats/blob/master/src/kernels.yaml#L7) which fail to unpack with `tar-rs`.

Example: http://mirrors.kernel.org/pub/linux/kernel/v1.1/linux-1.1.23.tar.gz

This works with gnu tar, but fails with `failed to read entire block` when trying to read the last header.

This patch makes the last header optional. I honestly have no idea if this is "correct" or not, but at the very least it seems to be conventional.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! I was trying to read around in GNU tar to see if I could find where this was handled there, but I was finding the code really difficult to navigate... Do you know if there's perhaps prior art somewhere of handling this sort of case? So long as at least one other library does it then it seems fine to me to include!

**Comment by udoprog** [comment]
I am also trying to read the gnu tar code right now. I'll poke back when I've figured it out.

**Comment by udoprog** [comment]
In gnu tar, this seems to be decided here:
http://git.savannah.gnu.org/cgit/tar.git/tree/src/list.c#n233

To me it looks like it stops at the first zeroed block unless `--ignore-zeros` is specified. This is also supported by the documentation for `--ignore-zeros`: https://www.gnu.org/software/tar/manual/html_node/Ignore-Zeros.html

I'm a bit confused by the logic of the warning message. But it does decode the last zero block, and it's clear to me though that it's not a fatal error in case the end of the archive returns `HEADER_END_OF_FILE` there.

**Comment by alexcrichton** [comment]
It looks like [Python's implementation does something similar](https://github.com/python/cpython/blob/18ba4851aeb5dc5d6b4d22adc8d0555d0c2fdfe5/Lib/tarfile.py#L1496-L1497) too. Ok then! Perhaps an option could be added to this library as well to ignore blocks of zeros (default to false) and then the same logic of stop-on-a-block-of-zeros could be implemented?

File: tests/all.rs [hunk 1]
```diff
@@ -494,15 +565,15 @@ fn extracting_malformed_tar_null_blocks() {
 
     let data = t!(ar.into_inner());
     let mut ar = Archive::new(&data[..]);
-    assert!(ar.unpack(td.path()).is_err());
+    assert!(ar.unpack(td.path()).is_ok());
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply tests/all.rs]: Hm ok so I think the old behavior (really old at this point) was to just skip zero blocks. The PR switched to requiring two trailing zero blocks. The behavior of `tar`, though, was to halt on any null block (but not error)

I think therefore that this PR matches `tar` still. Mind confirming though that this makes sense?

[the remark this answers is not in the document]
Comment (udoprog) [reply tests/all.rs]: Sounds right. It also seems like this test is no longer valid then?
This PR aligns us with gnu tar, for which this does not appear to be an error even.

File: tests/all.rs [hunk 2]
```diff
@@ -494,15 +565,15 @@ fn extracting_malformed_tar_null_blocks() {
...
 }
 
 #[test]
 fn empty_filename() {
     let td = t!(TempDir::new("tar-rs"));
     let rdr = Cursor::new(tar!("empty_filename.tar"));
     let mut ar = Archive::new(rdr);
-    assert!(ar.unpack(td.path()).is_err());
+    assert!(ar.unpack(td.path()).is_ok());
```
[the remark this answers is not in the document]
Comment (udoprog) [reply tests/all.rs]: Note that we only allow EOF on header boundaries, so a header either has to be all zeros or _completely missing_ to be counted as the end of an archive.

File: tests/all.rs [hunk 3]
```diff
@@ -58,6 +102,33 @@ fn header_impls() {
     }
 }
 
+#[test]
+fn simple_missing_last_header() {
+    let mut ar = Archive::new(Cursor::new(tar!("simple_missing_last_header.tar")));
+    for entry in t!(ar.entries()) {
+        t!(entry);
+    }
+    let mut ar = Archive::new(Cursor::new(tar!("simple_missing_last_header.tar")));
```
[the remark this answers is not in the document]
Comment (udoprog) [reply tests/all.rs]: Copied [`simple`](https://github.com/alexcrichton/tar-rs/blob/master/tests/all.rs#L35) without thinking too much about it.

File: tests/all.rs [hunk 4]
```diff
@@ -58,6 +102,33 @@ fn header_impls() {
     }
 }
 
+#[test]
+fn simple_missing_last_header() {
```
[the remark this answers is not in the document]
Comment (udoprog) [reply tests/all.rs]: yes

File: tests/all.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (udoprog) [reply tests/all.rs]: I probably should have :P.

File: tests/all.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (udoprog) [reply tests/all.rs]: Ah, now I get it. Yeah sure. They do the same. Removed.
There's also one from before for `simple.tar`, which this one was copied from. Removed those as well.

**Comment by alexcrichton** [comment]
Looks great to me, thanks so much again!

**Pull request merged**